### PR TITLE
Add Type System Package

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

This introduces a new package (set of packages) for the Type System as part of the implementation of #418. The definition of the ontology types, their builders, and their validation logic, will be written in Rust. This Rust is then compiled to WASM and turned into NPM packages that can be consumed in Web or NodeJS contexts. This should allow us to have consistent definitions of the Type System elements which we can use within Blocks (a Type Editor block for example) as well as in backend components written in both Node and Rust.

It should be possible in the future to generate other bindings to these implementations, potentially Python. It may also be possible to auto-generate JSON schema from the struct definitions themselves, which means that we would be able to keep the Block Protocol spec in sync with our libraries and reference implementations.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202775849963301/f) _(internal)_

## 🔍 What does this change?

- Adds `packages/type-system`
- Adds a Rust crate for Type System definitions within `packages/type-system/rust`
  - Adds a `cargo-make` setup to manage the various build and testing scripts
  - Adds a `wasm-pack` configuration to turn a `wasm_bindgen` configured crate into an NPM package with TypeScript bindings
  - Adds definitions and validation logic for `BaseURI` and `VersionedUri` of Type System types
- Adds a TypeScript folder to contain the generated NPM packages, and
  - `packages/ts/integration` which contains Jest integration tests which tests the generated NPM packages

## 📜 Does this require a change to the docs?

- READMEs have been added for the package itself. We likely want to improve discoverability of this package, and add usage instructions for how to import and use the NPM packages.

## ⚠️ Known issues

- `wasm-pack` has a `bundler` target which could allow us to create an ES module. Unfortunately jest is unable to correctly load that when it's made, and nextJS and `create-block-app` both struggle with handling the included `.wasm` asset when bundling the ES module as they mess up the loading mechanism and path. Because of this the current PR generates two different packages that can be used in either Web or NodeJS contexts.
- When passing a `number` through on the TypeScript side to Rust, if you pass a non-integer it gets floored without erroring. (So if you pass `2.3` you end up with the version `2`). A TODO has been added

## 🐾 Next steps

- Add the types in the ontology
  - Data Types
  - Property Types
  - Entity Types
  - Link Types
- Add the build and tests to CI
- Figure out a strategy for returning and handling typed Errors
- Introduce a publishing flow for the packages
- Introduce a publishing flow for the crate
- Use the types within a block 
- Use the crate within the projects over on [hashintel/hash](https://github.com/hashintel/hash)

## 🛡 What tests cover this?

- Integration tests have been added to test the generated NodeJS package
- Rust unit tests _will_ be added to also check the crate
- There is a setup for WASM headless browser tests but these are currently not utilised, a testing strategy will need to be developed

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Follow the READMEs

## 📹 Demo

image
Pretty boring demo but this is a VersionedUri being created within a Block

### CODE REVIEW

File: packages/type-system/ts/test-block/src/app.tsx
Comment (Alfred-Mountfield) [inline packages/type-system/ts/test-block/src/app.tsx]: This shows an example of the library working within a block. I deleted it as I assumed we don't want a random stray block sitting around without any automated testing as part of it.

File: packages/type-system/rust/src/ontology/uri/mod.rs
Comment (Alfred-Mountfield) [inline packages/type-system/rust/src/ontology/uri/mod.rs:58]: At the moment I'm swallowing all of the [useful errors](https://docs.rs/url/latest/url/enum.ParseError.html) from the `Url` crate. I thought it was out of scope of this PR to manage how to expose all of those to TS and how we'd potentially want to handle them.

File: packages/type-system/rust/src/ontology/uri/native.rs
Comment (Alfred-Mountfield) [inline packages/type-system/rust/src/ontology/uri/native.rs:7]: See comment above about swallowing [url errors](https://docs.rs/url/latest/url/enum.ParseError.html)

File: packages/type-system/rust/src/ontology/uri/native.rs
Comment (Alfred-Mountfield) [inline packages/type-system/rust/src/ontology/uri/native.rs:10]: This is quite strange behaviour but seems somewhat consistent with the reference implementation of the actual [URL spec](https://url.spec.whatwg.org/). I'm not sure what we want to do here because it does helpfully remove stray slashes and standardises some URLs, but at the same time the example I gave here seems like something we'd probably want errors for (things that _are_ considered errors by the library are [here](https://docs.rs/url/latest/url/enum.ParseError.html)).

File: packages/type-system/rust/src/ontology/uri/wasm.rs
Comment (Alfred-Mountfield) [inline packages/type-system/rust/src/ontology/uri/wasm.rs:11]: See comments above

File: packages/type-system/rust/tests/web.rs
Comment (Alfred-Mountfield) [inline packages/type-system/rust/tests/web.rs:10]: So `wasm-pack` has this nice feature where it will automatically download a driver for a browser on your system and will compile the module into WASM and actually run these tests (written normally in Rust) inside your browser. That seems quite helpful but I think it basically runs it inside WASM, so it's not testing the bindings, which is why I ended up going with the Jest setup. Perhaps there are still some tests we'd like to add here though.

File: packages/type-system/ts/README.md
Comment (Alfred-Mountfield) [inline packages/type-system/ts/README.md:4]: I'm not sure if we want to commit these to version control

File: packages/type-system/ts/integration/src/uri.test.ts
Comment (Alfred-Mountfield) [inline packages/type-system/ts/integration/src/uri.test.ts:66]: Just highlighting this as a current limitation

File: packages/type-system/rust/build-wasm.py
Comment (Alfred-Mountfield) [inline packages/type-system/rust/build-wasm.py:13]: There aren't great configuration options for the generated `package.json`s and if we're publishing to NPM I assumed we'd want to patch in some content. I wrote this script to setup the infra to do so but I don't really know what we want to patch in.

File: packages/type-system/ts/integration/package.json
Comment (Alfred-Mountfield) [inline packages/type-system/ts/integration/package.json:10]: @kachkaev CI seems to be failing because of these
```
error Couldn't find package "@blockprotocol/type-system-node@0.0.0" required by "@blockprotocol/type-system-integration@0.0.0" on the "npm" registry.
```

Should I swap to the `file:/..` syntax? Or is there something else we can do

File: packages/type-system/ts/README.md
Comment (Alfred-Mountfield) [reply packages/type-system/ts/README.md:4]: (it's worth noting that these are built packages, not plain source code that needs transpiling)

If we don't commit them then I think we'll need to figure out a publishing mechanism before we're able to use these easily in a block. If we do commit it then I think it should be simple to install them by just referencing their link in git

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/tests/uri.test.ts:12]: We can try using `test.each` / `it.each` here instead

File: packages/type-system/ts/integration/package.json
Comment (Alfred-Mountfield) [reply packages/type-system/ts/integration/package.json]: This has been solved by adding the packages to version control

File: packages/type-system/ts/README.md
Comment (Alfred-Mountfield) [reply packages/type-system/ts/README.md]: We've decided it makes sense to add them to version control. When we come to adding CI checks we'll need to check for a diff between the generated content and the committed one.

File: crates/type-system/Cargo.toml
Comment (Alfred-Mountfield) [inline crates/type-system/Cargo.toml:19]: Turns out URL is adding a _**lot**_ (~300-400kb) to the size of the WASM binary. This is a [known issue](https://github.com/servo/rust-url/issues/557) but I don't think they're planning on addressing it.

My current thinking is we go with it until it seems like the size is a problem, at which point we could write a lightweight URL thing that does what we need from it, or we could conditionally [utilise the browser's implementation](https://docs.rs/web-sys/0.3.45/web_sys/struct.Url.html) of `url` in a wasm environment.

Either way we have options and I don't think we need to worry right now.

File: crates/type-system/build-wasm.ts
Comment (kachkaev) [inline crates/type-system/build-wasm.ts:1]: Moving this file to `crates/type-system/scripts/build-wasm.ts` should fix most of ESLint errors. You can also run `yarn fix` locally to address a few. We can look at the remaining ones afterwards.

File: .eslintignore
Comment (kachkaev) [inline .eslintignore:66]: We can copy this to `.prettierignore` too

File: crates/type-system/.gitignore
Comment (kachkaev) [inline crates/type-system/.gitignore:6]: Can we hoist this to top-level `.gitignore`?

File: crates/type-system/README.md
Comment (kachkaev) [inline crates/type-system/README.md:6]: Try running `yarn fix:markdownlint` and `yarn fix:prettier` from repo root to fix some formatting issues in this README.

File: crates/type-system/.gitignore
Comment (Alfred-Mountfield) [reply crates/type-system/.gitignore]: That's been done https://github.com/blockprotocol/blockprotocol/pull/503/commits/9d0cf95cc83e95a1c2ed20e4540796c417207b3b

Let me know if that's not how you intended it to be

File: .eslintignore
Comment (Alfred-Mountfield) [reply .eslintignore:66]: Updated the ignore files 
https://github.com/blockprotocol/blockprotocol/pull/503/commits/ac445d6c1b9193d089b951d4da434febc60011fd

File: crates/type-system/build-wasm.ts
Comment (Alfred-Mountfield) [reply crates/type-system/build-wasm.ts]: That was done https://github.com/blockprotocol/blockprotocol/pull/503/commits/3891af587f3efc63e1fa992637fe8056c3f0fb62

Thanks :)

File: crates/type-system/README.md
Comment (Alfred-Mountfield) [reply crates/type-system/README.md:7]: Was holding off until I figured out the appropriate ignores.

Applied the fix https://github.com/blockprotocol/blockprotocol/pull/503/commits/6dbd20e25655eb5eaa812509cb81a2ba38e36126

(and removed unnecessary README)

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/tests/uri.test.ts:1]: Might be fixable via

```suggestion
import { BaseUri, VersionedUri } from "..";
```

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/tests/uri.test.ts:3]: `'describe' is not defined` can be fixable via top-level `.eslntrc`. Try finding `overrides` and adding this item there:

```
    {
      files: ["packages/**/*.test.{j,t}s{x,}"],
      env: {
        jest: true,
        node: true,
      },
    },
```

Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/tests/uri.test.ts:3]: Thanks! That was stumping me

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/tests/uri.test.ts]: As discussed offline, this was a misguided attempt to test the `package.json`.
Changed in https://github.com/blockprotocol/blockprotocol/pull/503/commits/d93057ed98481477f6feeee5c1072c988dc578ca

File: packages/type-system/rust/build-wasm.py
Comment (Alfred-Mountfield) [reply packages/type-system/rust/build-wasm.py]: This has been made redundant

File: packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts
Comment (thehabbos007) [inline packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: Should we add `.gitattributes` for the auto-generated files similarly to what we did for #928 (this goes for the other package as well)?

Comment (kachkaev) [reply packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: We can yes. What do we get from doing so in this case?

File: crates/type-system/Cargo.toml
Comment (TimDiekmann) [inline crates/type-system/Cargo.toml:33]: Given the size of the binary is it worth pulling an external allocator for this?

File: crates/type-system/rustfmt.toml
Comment (TimDiekmann) [inline crates/type-system/rustfmt.toml:1]: Is this a 1:1 copy from hashintel/hash?

File: crates/type-system/src/lib.rs
Comment (TimDiekmann) [inline crates/type-system/src/lib.rs:1]: Do we want to pull in the same lints as for the HASH backend?

File: crates/type-system/src/lib.rs
Comment (TimDiekmann) [inline crates/type-system/src/lib.rs:6]: Not really a fan of this comment as this is exactly what is written in the code below

File: crates/type-system/Cargo.toml
Comment (TimDiekmann) [inline crates/type-system/Cargo.toml:20]: I think it's possible to only pull it for the wasm target:

```toml
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = { version = "0.2.63", features = ["serde-serialize"] }
```

File: crates/type-system/src/ontology/uri/wasm.rs
Comment (TimDiekmann) [inline crates/type-system/src/ontology/uri/wasm.rs:21]: Is there something like `toCamelCase`?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: crates/type-system/rustfmt.toml
Comment (Alfred-Mountfield) [reply crates/type-system/rustfmt.toml:1]: It is

File: crates/type-system/Cargo.toml
Comment (Alfred-Mountfield) [reply crates/type-system/Cargo.toml:33]: I think so, yes. We're very likely to address the binary size problems from `url` and generally speaking a lot of people seem to use `wee_alloc` for WASM so I'm happy to stick with convention.

File: crates/type-system/src/lib.rs
Comment (Alfred-Mountfield) [reply crates/type-system/src/lib.rs:6]: Happy to delete it, the comment was from the boilerplate when the project was generated

File: crates/type-system/src/ontology/uri/wasm.rs
Comment (Alfred-Mountfield) [reply crates/type-system/src/ontology/uri/wasm.rs:21]: not that I've found https://github.com/rustwasm/wasm-bindgen/issues/1818

File: crates/type-system/src/lib.rs
Comment (Alfred-Mountfield) [reply crates/type-system/src/lib.rs:1]: Good idea, have brought most of them in, we can remove if they cause problems

File: crates/type-system/src/lib.rs
Comment (Alfred-Mountfield) [reply crates/type-system/src/lib.rs]: Have deleted

File: crates/type-system/Cargo.toml
Comment (Alfred-Mountfield) [reply crates/type-system/Cargo.toml]: I've moved most of the wasm stuff under that heading

File: packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts
Comment (thehabbos007) [reply packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: They'll be suppressed in diffs and excluded from the language statistics on the GitHub repo

**Review by thehabbos007 — COMMENTED** [review]
I don't have much to add to the review considering the upcoming PR to change binding code. The package output looks nice. I suppose the wildcard in the root `package.json` takes care of adding it to the Yarn workspace.

What is the plan for the HASH Graph backend, by the way? Will it depend on this type-system crate on a specific git revision, or go through crates.io?

Otherwise build/test passes fine locally here. I'd wait for Alex' final approval :+1:

File: packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: I'm happy to add this if you also think it's a good idea @kachkaev

**Comment by Alfred-Mountfield** [comment]
> What is the plan for the HASH Graph backend, by the way? Will it depend on this type-system crate on a specific git revision

At the moment I think during dev lets just target it via git, in the longer run we should look at publishing this crate (and the npm packages)

File: packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts
Comment (kachkaev) [reply packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: Yep, let's do that. Why not?

File: crates/type-system/scripts/build-wasm.ts
Comment (kachkaev) [inline crates/type-system/scripts/build-wasm.ts:2]: We can use `execa` instead, which has a slightly more convenient API. No need to do this now, this is just an idea.

File: packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/src/index_bg.wasm.d.ts:1]: That's been [done](https://github.com/blockprotocol/blockprotocol/pull/503/files/15ef69e4a87fea093bae318405001dab2f02f9c4..994b6b238ed95057452f443ec3272b14e0a5af49)

I assume that's what you had in mind @thehabbos007?

File: crates/type-system/scripts/build-wasm.ts
Comment (kachkaev) [inline crates/type-system/scripts/build-wasm.ts:39]: Potential future improvement: might be worth avoiding `__dirname` in ESM modules. It's a built-in constant, which may look confusing (are we in CommonJS? 🤔)

We rarely  need script dir name, so can go for something like:

```ts
export const packageRootDirPath = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "..",
);
```

(copied from some other file)

File: crates/type-system/scripts/build-wasm.ts
Comment (kachkaev) [inline crates/type-system/scripts/build-wasm.ts:80]: Potential future enhancements:

```
const script = async () => {
  // ... 
};

await script();
```

(see other script files)

We can also switch to async fs functions for consistency with other code. Using sync functions is OK in scripts, but is undesired in API handlers etc. Using one style everywhere is simpler from the cognitive POV. There are fewer ifs in the rules we follow.

File: packages/@blockprotocol/type-system-node/.gitattributes
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/.gitattributes:1]: Let's avoid `.gitattributes` in subfolders to make repo config more transparent and easier to manage (at least for now). One source of truth helps simplify the set of rules and see the whole picture at a glance. If any folder can configure itself, that's quite a lot of extra scanning work for developers.

File: packages/@blockprotocol/type-system-node/package.json
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/package.json:3]: If we are ready to publish these packages, I'd version them as `0.0.1`. We can later use Changesets to prepare new versions.

**Review by kachkaev — COMMENTED** [review]
Looking good! Let's tweak `.gitattributes` and version numbers and :shipit:!

**Comment by kachkaev** [comment]
One more thing: [Broken `yarn lint:tsc`](https://github.com/blockprotocol/blockprotocol/runs/7920802267?check_suite_focus=true#step:9:54) should be fixable by adding `tsconfig.json` to `type-system-web`.

File: packages/@blockprotocol/type-system-node/package.json
Comment (kachkaev) [reply packages/@blockprotocol/type-system-node/package.json:3]: If we want to try merging packages before releasing, we can keep 0.0.0 for now.

File: crates/type-system/scripts/build-wasm.ts
Comment (Alfred-Mountfield) [reply crates/type-system/scripts/build-wasm.ts:80]: Going to avoid committing this one. I appreciate the argument about consistency but I don't think it's worth the dev effort to swap to an async API here when I need the script to behave synchronously anyway

File: packages/@blockprotocol/type-system-node/package.json
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/package.json:3]: I don't think they're worth publishing in their current state at all as is. I think there'll be a separate PR like "Prep type-system packages for publishing" or something in the future. (I suspect when we've implemented the 4 ontology types and it's not just a convoluted, oversized URI validation library 😁) We can try and revisit merging them at that point, as well as making sure we're happy with the package.json's, descriptions, etc.

File: packages/@blockprotocol/type-system-node/.gitattributes
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/.gitattributes]: I've combined and moved to the top-level in a new file as we didn't seem to have one

File: crates/type-system/scripts/build-wasm.ts
Comment (Alfred-Mountfield) [reply crates/type-system/scripts/build-wasm.ts:3]: Have captured as a @todo

**Review by TimDiekmann — DISMISSED** [review]
Rust code looks good!

Looking forward also to use this `Makefile.toml` for the backend 🎉

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/tests/uri.test.ts:1]: It' usually `test` instead of `tests`, but we can keep the folder name as is for now.

**Comment by kachkaev** [comment]
@TimDiekmann / @thehabbos007 do you have any outstanding suggestions? Feel free to approve if not 🙂

File: packages/@blockprotocol/type-system-node/tests/uri.test.ts
Comment (Alfred-Mountfield) [reply packages/@blockprotocol/type-system-node/tests/uri.test.ts:1]: I'll fix on the next PR, thanks :)

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**
